dali/pipeline/operator/batch_size_provider.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATOR_BATCH_SIZE_PROVIDER_H_
+#define DALI_PIPELINE_OPERATOR_BATCH_SIZE_PROVIDER_H_
+
+namespace dali {
+
+class BatchSizeProvider {
+  /*
+   * Typically, you'll need to inherit from this interface virtually:
+   *
+   * template<typename Backend>
+   * class MyOperatorBsp : public OperatorBase<Backend>, virtual public BatchSizeProvider {};
+   */
+ public:
+  /**
+   * Returns next batch size.
+   *
+   * Implementation shall assure, that it's possible to call NextBatchSize()
+   * multiple times for the same batch, before Advance() invocation.
+   */
+  virtual int NextBatchSize() = 0;
+
+  /**
+   * Advances to next batch.
+   *
+   * On the event, that there's no further data available, Advance() shall throw std::out_of_range
+   */
+  virtual void Advance() = 0;
```

Comment:
I thought about it, but from my perspective that's inaccurate. This function does not advance the "next batch size". It does not advance "batch size" either. It advances to the next batch, so it could be this way: `AdvanceToNextBatch`. But that "next batch" is not the "next batch" from the `NextBatchSize` function. Anyway, I believe simple `Advance` with proper documentation is accurate enough.